include/geode/geometry/aabb.h

```diff
@@ -0,0 +1,472 @@
+            const ACTION& action ) const;
+
+        template < class ACTION >
+        void bbox_intersect_recursive( const BoundingBox< dimension >& box,
+            index_t node_index,
+            index_t element_begin,
+            index_t element_end,
+            ACTION& action ) const;
+
+        template < class ACTION >
+        void self_intersect_recursive( index_t node_index1,
+            index_t element_begin1,
+            index_t element_end1,
+            index_t node_index2,
+            index_t element_begin2,
+            index_t element_end2,
+            ACTION& action ) const;
+
+        /*!
+         * @brief Gets an hint of the result
+         * @details Compute the result by approximating each bbox to its
+         * barycenter (using given distance evaluator).
+         * This result is then used to speed-up the computation by minimizing
+         * the distance computation between \p query and the real elements
+         * inside the bboxes
+         */
+        template < class EvalDistance >
+        std::tuple< index_t, Point< dimension >, double >
+            get_nearest_element_box_hint( const Point< dimension >& query,
+                const EvalDistance& eval_distance ) const
+        {
+            index_t box_begin{ 0 };
+            index_t box_end{ nb_bboxes() };
+            index_t node_index{ ROOT_INDEX };
+            while( !is_leaf( box_begin, box_end ) )
+            {
+                index_t box_middle;
+                index_t child_left;
+                index_t child_right;
+                get_recursive_iterators( node_index, box_begin, box_end,
+                    box_middle, child_left, child_right );
+                if( eval_distance(
+                        ( node( child_left ).min() + node( child_left ).max() )
+                            / 2.,
+                        query )
+                    < eval_distance( ( node( child_right ).min()
+                                         + node( child_right ).max() )
+                                         / 2.,
+                        query ) )
+                {
+                    box_end = box_middle;
+                    node_index = child_left;
+                }
+                else
+                {
+                    box_begin = box_middle;
+                    node_index = child_right;
+                }
+            }
+
+            const auto nearest_box = mapping_morton_[box_begin];
+            const auto nearest_point =
+                get_point_hint_from_box( tree_[box_begin], nearest_box );
+            const auto distance = eval_distance( query, nearest_point );
+            return std::make_tuple( nearest_box, nearest_point, distance );
+        }
+        /*!
+         * @brief Gets an element point from its box
+         * @details This function is used to get a result from the selected hint
+         * box
+         */
+        virtual Point< dimension > get_point_hint_from_box(
```

Comment:
This method allows a better distance hint approximation for classes that inherit from AABBTree. Because of this, we add a vtable to the class. 

My question: is the box center a good enough approximation to remove the vtable? Is [this](https://github.com/ringmesh/RINGMesh/blob/master/src/ringmesh/mesh/mesh_aabb.cpp#L213) approximation modification really important?

Reply:
the thing is that depending on the object that is in the box. It may not be relevant to only consider it center. The box may be very far from a cubix box.

Reply:
true